controller/utils.go

```diff
@@ -61,165 +60,182 @@ func SegmentAndReplacePods(
 		"scheduler": schedulerName,
 	})
 
+	ctx, cancel := context.WithDeadline(context.Background(), willTimeoutAt)
+	defer cancel()
+
+	inRollingUpdate := operationManager != nil
+
+	if inRollingUpdate {
+		go watchOperation(ctx, cancel, operationManager)
+	}
+
 	// segment pods in chunks
 	podChunks := segmentPods(pods, maxSurge)
 
 	for i, chunk := range podChunks {
 		l.Debugf("updating chunk %d: %v", i, names(chunk))
 
 		// replace chunk
-		timedout, canceled, errored := replacePodsAndWait(
-			l,
+		err := replacePodsAndWait(
+			ctx,
+			logger,
 			roomManager,
 			mr,
 			clientset,
 			db,
 			redisClient,
-			willTimeoutAt,
 			configYAML,
 			chunk,
 			scheduler,
-			operationManager,
-			clock,
+			inRollingUpdate,
 			goroutinePoolSize,
 		)
 
-		if timedout {
-			timeoutErr = errors.New("timedout waiting rooms to be replaced, rolled back")
-			l.WithError(timeoutErr).Error("error replacing chunk of pods")
-			break
+		if err != nil {
+			l.WithError(err).Error("error replacing chunk of pods")
+			return nil, nil, err
 		}
+	}
 
-		if canceled {
-			cancelErr = errors.New("operation was canceled, rolled back")
-			l.WithError(cancelErr).Error("error replacing chunk of pods")
-			break
-		}
+	cancelReason := ctx.Err()
 
-		if errored != nil {
-			err = errored
-			l.WithError(errored).Error("error replacing chunk of pods")
-			break
-		}
+	if cancelReason == context.DeadlineExceeded {
+		timeoutErr = errors.New("timedout waiting rooms to be replaced, rolled back")
+		l.WithError(timeoutErr).Error("operation timed out while replacing chunk of pods")
+	} else if cancelReason == context.Canceled {
+		cancelErr = errors.New("operation was canceled, rolled back")
+		l.WithError(cancelErr).Error("operation canceled while error replacing chunk of pods")
 	}
 
 	return timeoutErr, cancelErr, err
 }
 
+func watchOperation(ctx context.Context, cancel func(), operationManager *models.OperationManager) {
+	canceled, err := operationManager.WasCanceled()
+	if canceled && err  == nil {
+		cancel()
+		return
+	}
+
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			canceled, err := operationManager.WasCanceled()
+			if err != nil {
+				continue
```

Comment:
In this case where `WasCanceled` returns an error (it can only be a Redis problem). We have two choices here:
* This problem could be resolved at the next tick but if not we'll keep this running until the timeout is reached (not sure if any other operation would have some problem using redis also, but I think we can't assume that);
* Cancel the operation since we actually don't know if Redis has the key or not.

My concern here is that in both cases we can have some "keys leak". What do you guys think?

Not sure if we what to solve this in this PR but, I would recommend adding a `warn` log here so we are aware that this Redis failures are happening here.

Reply:
A log helps, we can also add a maximum amount of times, and then it fails with an error maybe?

Reply:
Failing will not help here I guess as well.

Reply:
I think that having a failure helps, it would be a way of us taking control and fixing things manually maybe, without waiting for the timeout which is really long. Maybe adding a metric here so that we can have an alarm. A log is probably enough however.

Reply:
After talking with corado on slack I think just logging would be enough. If this error happened once only in this place, it would be ok on the next try and if it a generalized redis problem the operation would fail creating or deleting pods and would cancel anyway.